### PR TITLE
Add support for OSS protocol version 2

### PR DESCRIPTION
**Pull request by adutra**
TODO

- [x] Add unit tests for newly-added functions
- [x] Test manually with Java driver 3.x to ensure that v2 is well implemented

### CODE REVIEW

File: client/main_test.go [hunk 1]
```diff
@@ -79,27 +81,35 @@ func createCodecs() {
...
+	var fixed = func(clientId int, version primitive.ProtocolVersion) int16 {
 		if int16(clientId) == client.ManagedStreamId {
 			panic("stream id 0")
 		}
 		return int16(clientId)
 	}
-	counter := uint32(0)
-	var incremental = func(clientId int) int16 {
+	counter := uint32(1)
+	var incremental = func(clientId int, version primitive.ProtocolVersion) int16 {
```
Comment (joao-r-reis) [inline client/main_test.go:94]: `clientId` isn't used anymore? Also, why don't you use a buffered channel that is pre-populated with the stream ids?

**Review by joao-r-reis — APPROVED** [review]
Most of this PR diff is related to renaming functions/variables, comments and other small changes (unrelated to the OSS protocol v2) so it's a bit hard to pinpoint exactly what changes are related to the protocol v2 support but the new `primitives` functions `ReadStreamid` and `WriteStreamid` are a good idea, I like it.

I also like having a `NewFrame` function and determining whether it is a response or not via the message type.

💯 to all the code comments that were added.

**Comment by adutra** [comment]
> Most of this PR diff is related to renaming functions/variables, comments and other small changes (unrelated to the OSS protocol v2) so it's a bit hard to pinpoint exactly what changes are related to the protocol v2 support but the new `primitives` functions `ReadStreamid` and `WriteStreamid` are a good idea, I like it.
> 
> I also like having a `NewFrame` function and determining whether it is a response or not via the message type.
> 
> 💯 to all the code comments that were added.

Sorry I need to keep my PRs more focused indeed. Thanks for reviewing nonetheless.

I added much more detailed tests to the client-server API, because I think we need to have more confidence that encoding/decoding is actually working against a real C* node. If you look into `TestRemoteServerNoAuth` and alike, they now test every possible message. I just validated manually that the tests passes for v2 with C* 2.1, so I'm fairly confident at this point that v2 is fully implemented.

**Comment by joao-r-reis** [comment]
> 
> 
> > Most of this PR diff is related to renaming functions/variables, comments and other small changes (unrelated to the OSS protocol v2) so it's a bit hard to pinpoint exactly what changes are related to the protocol v2 support but the new `primitives` functions `ReadStreamid` and `WriteStreamid` are a good idea, I like it.
> > I also like having a `NewFrame` function and determining whether it is a response or not via the message type.
> > 💯 to all the code comments that were added.
> 
> Sorry I need to keep my PRs more focused indeed. Thanks for reviewing nonetheless.
> 
> I added much more detailed tests to the client-server API, because I think we need to have more confidence that encoding/decoding is actually working against a real C* node. If you look into `TestRemoteServerNoAuth` and alike, they now test every possible message. I just validated manually that the tests passes for v2 with C* 2.1, so I'm fairly confident at this point that v2 is fully implemented.

Awesome stuff 👍 

I'm ok with this type of PRs, sometimes you're modifying the code and you find small changes that you'd like to implement and it is faster to just do it on the stop. My suggestion would be to at least try and separate those changes in separate commits so the reviewer can use the github UI to exclude/include individual commits from the diff which makes it easier to review.

Of course I say this but I do the exact same thing on my PRs and I don't always separate those changes in separate commits like I should 😅

**Pull request merged**